src/test/java/com/cloudbees/jenkins/support/impl/SlaveCommandStatisticsTest.java

```diff
@@ -77,4 +78,44 @@ public Void invoke(File f, VirtualChannel channel) throws IOException, Interrupt
         }
     }
 
+    @Test @Issue("JENKINS-58528")
+    public void statisticsAreRotatedWithComputers() throws Exception {
+        SlaveCommandStatistics scs = ExtensionList.lookupSingleton(SlaveCommandStatistics.class);
+
+        SlaveCommandStatistics.MAX_STATS_SIZE = 0;
+        assertThat(scs.getStatistics().size(), equalTo(0));
+        DumbSlave s0 = r.createOnlineSlave();
+        assertThat(scs.getStatistics().size(), equalTo(1));
+        r.jenkins.removeNode(s0);
+        assertThat(scs.getStatistics().size(), equalTo(0));
+
+        SlaveCommandStatistics.MAX_STATS_SIZE = 1;
+        DumbSlave s1 = r.createOnlineSlave();
+        DumbSlave s2 = r.createOnlineSlave();
+        DumbSlave s3 = r.createOnlineSlave();
+        assertThat(scs.getStatistics().size(), equalTo(3));
+        r.jenkins.removeNode(s1);
+        assertThat(scs.getStatistics().size(), equalTo(3));
+        r.jenkins.removeNode(s2);
+        assertThat(scs.getStatistics().size(), equalTo(2));
+        r.jenkins.removeNode(s3);
+        assertThat(scs.getStatistics().size(), equalTo(1));
+        assertThat("Latest preserved", scs.getStatistics().keySet(), contains(s3.getNodeName()));
+
+        SlaveCommandStatistics.MAX_STATS_SIZE = 3;
+        s0 = r.createOnlineSlave();
+        s1 = r.createOnlineSlave();
+        s2 = r.createOnlineSlave();
+        s3 = r.createOnlineSlave();
+        assertThat(scs.getStatistics().size(), equalTo(4));
+        r.jenkins.removeNode(s0);
+        assertThat(scs.getStatistics().size(), equalTo(4));
+        r.jenkins.removeNode(s1);
+        assertThat(scs.getStatistics().size(), equalTo(4));
+        r.jenkins.removeNode(s2);
+        assertThat(scs.getStatistics().size(), equalTo(4));
+        r.jenkins.removeNode(s3);
+        assertThat(scs.getStatistics().size(), equalTo(3));
+        assertThat("Latest preserved", scs.getStatistics().keySet(), contains(s3.getNodeName(), s2.getNodeName(), s1.getNodeName()));
```

Comment:
Is this checking the order? Or only that they are contained?

Reply:
yep, `containsInAnyOrder` does the other thing.

Reply:
Shouldn't this have the opposite order then? The ones that are removed first are the ones that should be in the result first.

Reply:
Note that current implementation has no time based ordering. Given the only thing we do with the data is we dump them into the bundle where they will be consumed after extraction in different order (alphabetical or FS specific one), I do not see much point in doing that. What is important here is those are the _most recent_ statistics we keep.

Reply:
I'm not worried about the order itself (though I don't understand why it's this one) but by the fact that the wrong ones could be removed. But that case seems to be covered by the fact that s0 is not present.

Reply:
The chosen order is given by implementation specifics of `LinkedHashMap` where popping from the end would have linear complexity while from the front it has constant: https://github.com/jenkinsci/support-core-plugin/pull/180/files#diff-f2444e1194b4e202ab31a9f3e90e9444R215

I was concerned about the same thing and I believe the tests are covering the fact it is the latest elements that are present after rotation, as you said.